### PR TITLE
[oneDNN] BatchMatMulV2+Mul+AddV2 fusion with oneDNN CPU backend

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR does a fusion of BatchMatMulV2 + Mul + AddV2 with oneDNN CPU backend using grappler remapper optimizer. Such pattern is common in Transformer based language models. It improves inference performance for higher batch sizes.

### CODE REVIEW

**Comment by mdfaijul** [comment]
@gbaned Resolved conflicts, thanks!

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h [hunk 1]
```diff
@@ -579,19 +587,22 @@ class MklMatMulPrimitive : public MklPrimitive {
 
   ~MklMatMulPrimitive() {}
 
-  void Execute(const T* a_data, const T* b_data, T* c_data,
+  void Execute(const Tlhs* a_data, const Trhs* b_data, Toutput* c_data,
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h:590]: Let's consolidate the implementation by making this function call
```Execute(a_data, b_data, c_data, nullptr, nullptr, stream);```
 instead.

**Review by penpornk — CHANGES_REQUESTED** [review]
Overall changes look okay to me. Adding @ezhulenev for remapper comments.

Also, if we are going to have more MKL-specific remapper matches, let's refactor all MKL remaps into their own file, e.g., mkl_remapper.cc (or onednn_remapper.cc) after TF 2.8 branch cut.

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 2]
```diff
@@ -2580,6 +2732,18 @@ Status Remapper::Optimize(Cluster* cluster, const GrapplerItem& item,
             &nodes_to_delete, is_gelu_approximate));
         continue;
       }
+
+      // Remap BatchMatMul+Mul+AddV2 into the _FusedBatchMatMul.
+      matched_nodes_map.clear();
+      remove_node_indices.clear();
+      if (FindFusedBatchMatMul(&ctx, i, &matched_nodes_map,
+                               &remove_node_indices)) {
+        string name = ctx.graph_view.GetNode(i)->GetName();
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Done.

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 3]
```diff
@@ -1475,6 +1476,99 @@ bool FindTensorToHashBucket(const RemapperContext& ctx, int node_index,
...
+        graph_matcher.GetMatchedNodes(fusion_pattern2, ctx->nodes_to_preserve,
+                                      ctx->graph_view.GetNode(node_index),
+                                      matched_nodes_map, remove_node_indices);
+  }
+
+  // OneDNN is not optimized for all shapes with regard to binary-post ops
+  // fusion. Allow limited cases only for now that are optimized, (i)
+  // multiplicand is scalar, (ii) BatchMatmulV2 output is 4D tensor, and (iii)
+  // addend is 4D tensor with second dim_size = 1.
+  if (found_op_type_match) {
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Done

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 4]
```diff
@@ -2360,6 +2466,52 @@ Status AddTensorToHashBucketNode(RemapperContext* ctx,
...
+  for (const auto& node_idx : remove_node_indices) {
+    (*nodes_to_delete)[node_idx] = true;
+  }
+  return Status::OK();
+}
+
+bool IsConv2DOrMatMul(const NodeDef& node) {
+  return IsConv2D(node) || IsMatMul(node);
+}
+
```
[the remark this answers is not in the document]
Comment (mdfaijul) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Done

File: tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h (hunk 1 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h]: Done

**Comment by mdfaijul** [comment]
@penpornk Thank you very much for the comments. I have addressed those. Please check.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Comment by mdfaijul** [comment]
@penpornk Could you please let me know what is failed as shown `Google internal checks FAILED`?

**Pull request merged**